v3/impl.h

```diff
@@ -221,13 +225,24 @@ struct wide_integer<Bits, Signed>::_impl {
             r = -r;
         }
 
-        size_t count = r / std::numeric_limits<uint64_t>::max();
+        size_t count = r / max_uint;
         self = count;
-        self *= std::numeric_limits<uint64_t>::max();
+        self *= max_uint;
         long double to_diff = count;
-        to_diff *= std::numeric_limits<uint64_t>::max();
-
-        self += to_Integral(r - to_diff);
+        to_diff *= max_uint;
+
+        /// There are values in int64 that have more than 52 significant bits (in terms of double
+        /// representation). Such values, being promoted to double, are rounded up or down. If they are rounded up,
+        /// the result may not fit in 64 bits.
+        /// The example of such a number is 9.22337e+18.
+        /// As to_Integral does a static_cast to int64_t, it may result in UB.
+        if (static_cast<__int128_t>(r - to_diff) > static_cast<__int128_t>(max_int))
```

Comment:
__int128_t is an unportable extension. Will `(r - to_diff) > static_cast<long double>(max_int)` do the job? You may add a static assert, that long double has enough bits

Reply:
@apolukhin Looks like the long double would do on most systems, and now there's a static assertion for the others.